filecoin-proofs/src/api/mod.rs

```diff
@@ -1,404 +1,120 @@
 
-    // Verification is cheap when parameters are cached,
-    // and it is never correct to return a proof which does not verify.
-    verify_seal(
-        porep_config,
-        comm_r,
-        comm_d,
-        prover_id,
-        sector_id,
-        ticket,
-        &buf,
-    )
-    .expect("post-seal verification sanity check failed");
-
-    Ok(SealOutput {
-        comm_r,
-        comm_d,
-        p_aux,
-        proof: buf,
-        comm_ps,
-        piece_inclusion_proofs,
-    })
-}
-
-/// Verifies the output of some previously-run seal operation.
-///
-pub fn verify_seal(
-    porep_config: PoRepConfig,
-    comm_r: Commitment,
-    comm_d: Commitment,
-    prover_id: ProverId,
-    sector_id: SectorId,
-    ticket: Ticket,
-    proof_vec: &[u8],
-) -> error::Result<bool> {
-    let sector_bytes = PaddedBytesAmount::from(porep_config);
+    let mut stack: Vec<PieceInfo> = vec![];
 
-    let comm_r = as_safe_commitment(&comm_r, "comm_r")?;
-    let comm_d = as_safe_commitment(&comm_d, "comm_d")?;
+    let reduce_inner = |stack: &mut Vec<PieceInfo>| {
+        if stack.is_empty() {
+            return false;
+        }
 
-    let replica_id =
-        generate_replica_id::<DefaultTreeHasher>(&prover_id, sector_id.into(), &ticket, comm_d);
+        let mut i = stack.len() - 1;
+        while i > 0 {
+            // println!("reduce_inner ({}, {}) {:?}", i, stack.len(), &stack);
+            let a = &stack[i - 1];
+            let b = &stack[i];
+
+            if a.size == b.size {
+                // println!("\tupdating");
+                let h = piece_hash(&a.commitment, &b.commitment);
+                // update a
+                stack[i - 1].commitment.copy_from_slice(h.as_ref());
+                stack[i - 1].size = UnpaddedBytesAmount(u64::from(stack[i - 1].size) * 2);
+                stack.pop(); // remove b
+
+                if i > 1 {
+                    i -= 2;
+                } else {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
 
-    let compound_setup_params = compound_proof::SetupParams {
-        vanilla_params: &setup_params(
-            PaddedBytesAmount::from(porep_config),
-            usize::from(PoRepProofPartitions::from(porep_config)),
-        ),
-        engine_params: &(*ENGINE_PARAMS),
-        partitions: Some(usize::from(PoRepProofPartitions::from(porep_config))),
+        stack.len() > 1
     };
 
-    let compound_public_params: compound_proof::PublicParams<
-        '_,
-        Bls12,
-        StackedDrg<'_, DefaultTreeHasher>,
-    > = StackedCompound::setup(&compound_setup_params)?;
-
-    let public_inputs = stacked::PublicInputs::<<DefaultTreeHasher as Hasher>::Domain> {
-        replica_id,
-        tau: Some(Tau { comm_r, comm_d }),
-        seed: None,
-        k: None,
+    let reduce = |stack: &mut Vec<PieceInfo>| {
+        // println!("\n---- reduce ----\n");
+        while reduce_inner(stack) {}
```

Comment:
This looks good. I don't think it yet handles subtree-alignment  padding, though.

In order to ensure that pieces are always subtree-aligned, the packing process may need to add a variable number of zero-padded bytes. These bytes don't correspond to actual pieces, so they won't appear in the list of `PieceInfo`s that are directly supplied by the chain when verifying.

The cleanest way to handle this is probably to walk the supplied `PieceInfo`s and insert these 'padding pieces' explicitly. It might make sense to precompute (or at least cache after first computation) the `CommP`s for the various sizes of padding which might be needed — since otherwise the cost has to be paid every time by all verifiers.

Once the shadow padding pieces have been inserted, a variation of the approach implemented here should work fine. While there is probably a way to compute this padding on the fly while reducing, I think it becomes complicated — so we should probably just directly apply the padding algorithm to the supplied `CommP`s. The variation required is that if the top (last) item on the stack is larger than the next, then it may need to be removed and held by pushing onto a second stack, while the remaining items are reduced until a long-enough piece has been produced. Then the new top of stack and the held item can be hashed as before. (The example below should exercise this.)

We should definitely add tests exercising this — ideally with randomized piece sizes inserted into a sector until sealing is triggered.

Here's a single example of pieces which I think the current implementation won't handle: A=1, B=4,C=2, D=8

These should be packed as follows:

[{(A0 00) (BB BB)} {(CC 00) (00 00)}] [{(DD DD) (DD DD)} {(00 00) (00 00)}]

This implies the following modified piece ordering (padding pieces labeled as P):

A(1), P(1), P(2), B(4), C(2), P(2), P(4), D(8), P(8)

Verification would proceed as follows:
- Compare lengths of P(8), D(8)
- Equal so Hash(D||P) to get DP(16)
- Compare lengths of DP(16) , P(4).
- DP is greater, so remove from stack S1 and push to hold on S2.
- Compare lengths of P(4), P(2).
- P(4) is greater, so remove from stack S1 and push to hold on S2.
- Compare lengths of C(2), P(2).
- Equal so Hash(C||P) to get CP(4).
- Compare lengths of CP(4), P(4) [top of S2]
- Equal, so Hash(CP(4)||P(4)) to get CPP(8) [and pop S2]
- Compare lengths of CPP(8), DP(16)
- DP is greater, so keep holding
- Compare lengths of CPP(8), B(4).
- CPP(8) is greater, so push on S2.
- Compare lengths of B(4), P(2).
- B(4) is greater, so push on S2.
- Compare lengths of P(2), P(1).
- P(2) is greater, so push on S2.
- Compare lengths of P(1), A(1).
- Equal, so Hash(A(1)||P(1)) to get AP(2).
- Compare lengths of AP(2), P(2) [top of S2]
- Equal so Hash(AP(2)||P(2)) to get APP(4) [and pop S2]
- Compare lengths of APP(4) and B(4) [top of S2]
- Equal so Hash(APP(4), B(4)) to get APPB(8) [and pop S2]
- Compare lengths of APPB(8) and CPP(8) [top of S2]
- Equal, so Hash (APPB(8)||CPP(8)) to get APPBCCPP(16) [and pop S2]
- Compare length of APPBCCPP(16) and DP(16)
- Equal, so Hash(APPBCCPP(16)||DP(16)) to get APPBCCPPDP(32) [and pop S2].
- Both stacks are empty, so APPBCCPPDP(32) is the result.
- Verify that the result equals CommD.

Note that APPBCCPPDP(32) is the expected result and has been constructed only by hashing equal-sized components.

Reply:
@porcuquine I used the logic of your description to devise an algorithm, going forwad and only using one stack + a current hash.

Reply:
I think what you did is equivalent. Using a 'stack' as a source of elements was just an artifact of going backward.

Reply:
Actually, I'm not so sure. Ignore my comments here and wait for me to submit review. Sorry for the noise.